rasterio/mask.py

```diff
@@ -3,132 +3,182 @@
 
-    if crop:
-        bounds_window = raster.window(*mask_bounds)
 
-        # Get the window with integer height
-        # and width that contains the bounds window.
-        out_window = bounds_window.round_shape(op='ceil')
-        height = int(out_window.height)
-        width = int(out_window.width)
+def mask(raster, shapes, all_touched=False, invert=False, nodata=None,
+         filled=True, crop=False, pad=False):
+    """Creates a masked or filled array using input shapes.  
+    Pixels are masked or set to nodata outside the input shapes, unless 
+    `invert` is `True`.
+    
+    If shapes do not overlap the raster and crop=True, an exception is 
+    raised.  Otherwise, a warning is raised.
 
-        out_shape = (raster.count, height, width)
-        out_transform = raster.window_transform(out_window)
+    Parameters
+    ----------
+    raster: rasterio RasterReader object
+        Raster to which the mask will be applied.
+    shapes: list of polygons
+        GeoJSON-like dict representation of polygons that will be used to 
+        create the mask.
+    all_touched: bool (opt)
+        Use all pixels touched by features. If False (default), use only
+        pixels whose center is within the polygon or that are selected by
+        Bresenham's line algorithm.
+    invert: bool (opt)
+        If True, mask will be `True` for pixels inside shapes and `False`
+        outside shapes.
+        False by default.
+    nodata: int or float (opt)
+        Value representing nodata within each raster band. If not set,
+        defaults to the nodata value for the input raster. If there is no
+        set nodata value for the raster, it defaults to 0.
+    filled: bool (opt)
+        If True, the pixels outside the features will be set to nodata.  
+        If False, the output array will contain the original pixel data, 
+        and only the mask will be based on shapes.  Defaults to True.
+    crop: bool (opt)
+        Whether to crop the raster to the extent of the shapes. Defaults to
+        False.
+    pad: bool (opt)
+        If True, the features will be padded in each direction by
+        one half of a pixel prior to cropping raster. Defaults to False.
 
-        logger.debug("Out window: %r", out_window)
-        logger.debug("Out transform: %r", out_transform)
+    Returns
+    -------
+    tuple
 
-    else:
-        out_window = None
-        out_shape = (raster.count, raster.height, raster.width)
-        out_transform = raster.transform
+        Two elements:
 
-    # Read the window of imagery.
-    out_image = raster.read(window=out_window, out_shape=out_shape,
-                            masked=True)
-    mask_shape = out_image.shape[1:]
+            masked : numpy ndarray or numpy.ma.MaskedArray
+                Data contained in raster after applying the mask (`filled` is 
+                `True`) and mask where all pixels are `True` outside shapes 
+                (`invert` is `False`).
```

Comment:
Given the complexity of the behavior (as you mentioned), I think better to err on the verbose side here. Something like:

"Data contained in the raster after applying the mask. If `filled` is True`, the return will be a `ndarray` in which masked areas are set to the `nodata` value. If `filled` is False, the return will be a `MaskedArray` in which `True` marks pixels outside shapes (or `False` if `invert` is `True`)."